File: N/1801-number-of-orders-in-the-backlog/main.go
package main

import (
	"container/heap"
)

type MinHeap []int

func (p MinHeap) Len() int {
	return len(p)
}

func (p MinHeap) Swap(i, j int) {
	p[i], p[j] = p[j], p[i]
}

func (p MinHeap) Less(i, j int) bool {
	return p[i] < p[j]
}

func (p *MinHeap) Push(c interface{}) {
	*p = append(*p, c.(int))
}
func (p *MinHeap) Pop() interface{} {
	n := len(*p)
	v := (*p)[n-1]
	*p = (*p)[0 : n-1]
	return v
}

type MaxHeap []int

func (p MaxHeap) Len() int {
	return len(p)
}

func (p MaxHeap) Swap(i, j int) {
	p[i], p[j] = p[j], p[i]
}

func (p MaxHeap) Less(i, j int) bool {
	return p[i] > p[j]
}

func (p *MaxHeap) Push(c interface{}) {
	*p = append(*p, c.(int))
}
func (p *MaxHeap) Pop() interface{} {
	n := len(*p)
	v := (*p)[n-1]
	*p = (*p)[0 : n-1]
	return v
}

func getNumberOfBacklogOrders(orders [][]int) int {
	sell := &MinHeap{}
	buy := &MaxHeap{}
	heap.Init(sell)
	heap.Init(buy)
	sm := map[int]int{}
	bm := map[int]int{}

	for _, order := range orders {
		amount := order[1]
		if order[2] == 0 {
			for amount > 0 && sell.Len() > 0 {
				p := heap.Pop(sell).(int)
				if p > order[0] {
					heap.Push(sell, p)
					break
				} else {
					diff := sm[p]
					if diff > amount {
						diff = amount
					}
					sm[p] -= diff
					amount -= diff
					if sm[p] == 0 {
						delete(sm, p)
					} else {
						heap.Push(sell, p)
					}
				}
			}
			if amount > 0 {
				_, ok := bm[order[0]]
				if !ok {
					heap.Push(buy, order[0])
				}
				bm[order[0]] += amount
			}
		} else if order[2] == 1 {
			for amount > 0 && buy.Len() > 0 {
				p := heap.Pop(buy).(int)
				// fmt.Println("sell", p, order[0])
				if p < order[0] {
					heap.Push(buy, p)
					break
				} else {
					diff := bm[p]
					if diff > amount {
						diff = amount
					}
					bm[p] -= diff
					amount -= diff
					if bm[p] == 0 {
						delete(bm, p)
					} else {
						heap.Push(buy, p)
					}
				}
			}
			if amount > 0 {
				_, ok := sm[order[0]]
				if !ok {
					heap.Push(sell, order[0])
				}
				sm[order[0]] += amount
			}
		}
		// fmt.Println(sell, buy, sm, bm)
	}

	ret := 0
	for _, v := range sm {
		v %= 1000000007
		ret += v
		ret %= 1000000007
	}
	for _, v := range bm {
		v %= 1000000007
		ret += v
		ret %= 1000000007
	}
	return ret
}

File: N/1801-number-of-orders-in-the-backlog/main.rs
struct Solution {}

use std::cmp::Reverse;
use std::collections::BinaryHeap;
use std::collections::HashMap;
impl Solution {
  pub fn get_number_of_backlog_orders(orders: Vec<Vec<i32>>) -> i32 {
    let mut sell: BinaryHeap<Reverse<i32>> = BinaryHeap::new();
    let mut buy: BinaryHeap<i32> = BinaryHeap::new();
    let mut samount: HashMap<i32, i32> = HashMap::new();
    let mut bamount: HashMap<i32, i32> = HashMap::new();

    for order in orders {
      let ot = order[2];
      match ot {
        0 => {
          let mut amount = order[1];
          while amount > 0 && sell.len() > 0 && sell.peek().unwrap().0 <= order[0] {
            let mut sa = samount.get_mut(&sell.peek().unwrap().0).unwrap();
            let mut diff = *sa;
            if diff > amount {
              diff = amount;
            }
            *sa -= diff;
            amount -= diff;
            if *sa == 0 {
              samount.remove(&sell.peek().unwrap().0);
              sell.pop();
            }
          }
          if amount > 0 {
            if bamount.contains_key(&order[0]) {
              let mut v = bamount.get_mut(&order[0]).unwrap();
              *v += amount;
            } else {
              buy.push(order[0]);
              bamount.insert(order[0], amount);
            }
          }
        }
        1 => {
          let mut amount = order[1];
          while amount > 0 && buy.len() > 0 && *buy.peek().unwrap() >= order[0] {
            let mut ba = bamount.get_mut(buy.peek().unwrap()).unwrap();
            let mut diff = *ba;
            if diff > amount {
              diff = amount;
            }
            *ba -= diff;
            amount -= diff;
            if *ba == 0 {
              bamount.remove(buy.peek().unwrap());
              buy.pop();
            }
          }
          if amount > 0 {
            if samount.contains_key(&order[0]) {
              let mut v = samount.get_mut(&order[0]).unwrap();
              *v += amount;
            } else {
              sell.push(Reverse(order[0]));
              samount.insert(order[0], amount);
            }
          }
        }
        _ => {}
      }
    }

    let mut ret: i64 = 0;
    for (_, mut v) in samount {
      v %= 1000000007;
      ret += v as i64;
      ret %= 1000000007
    }
    for (_, mut v) in bamount {
      v %= 1000000007;
      ret += v as i64;
      ret %= 1000000007
    }
    ret as i32
  }
}
